File: code/java/剑指offer31.java
class Solution {
    public boolean validateStackSequences(int[] pushed, int[] popped) {
        Stack<Integer> stack = new Stack<>();
        int push_i = 0;
        int pop_i = 0;
        while(push_i<pushed.length||pop_i<popped.length){
            if(stack.size()==0||stack.peek()!=popped[pop_i]){
                if(push_i>=pushed.length) return false;
                stack.push(pushed[push_i]);
                push_i++;
            }else{
                if(stack.size()>0){
                    stack.pop();
                    pop_i+=1;
                }else return false;
            }
        }
        return true;
    }
}

File: code/python/剑指offer31.py
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        push_i = 0
        pop_i = 0

        while push_i<len(pushed) or pop_i<len(popped):
            if len(stack)==0 or stack[-1]!=popped[pop_i]:
                if push_i>=len(pushed):
                    return False
                stack.append(pushed[push_i])
                push_i+=1
            else:
                if stack:
                    stack.pop()
                    pop_i+=1
                else:
                    return False
        return True
